File: JSDiffTests/JSDiffTests.m
//
//  JSDiffTests.m
//  JSDiff
//
//  Created by Piet Brauer on 09/04/16.
//  Copyright © 2016 nerdish by nature. All rights reserved.
//

#import <XCTest/XCTest.h>
#import "JSDiffTests-Swift.h"
#import <JSDiff/JSDiff.h>

@interface JSDiffTests : XCTestCase

@end

@implementation JSDiffTests

- (void)testSmallerLines {
    NSString *oldLine = @"<div class=\"{% if modulo == '6' %}flex-first{% endif %} sm-col sm-col-6 center\">";
    NSString *newLine = @"<div class=\"{% if modulo == '0' %}flex-first{% endif %} sm-col sm-col-4 center\">";
    NSMutableAttributedString *expectedOldLine = [[NSMutableAttributedString alloc] initWithString:oldLine attributes:@{NSBackgroundColorAttributeName: [UIColor deletedColor]}];
    [expectedOldLine addAttribute:NSBackgroundColorAttributeName value:[UIColor strongDeletedColor] range:NSMakeRange(29, 1)];
    [expectedOldLine addAttribute:NSBackgroundColorAttributeName value:[UIColor strongDeletedColor] range:NSMakeRange(70, 1)];
    NSMutableAttributedString *expectedNewLine = [[NSMutableAttributedString alloc] initWithString:newLine attributes:@{NSBackgroundColorAttributeName: [UIColor addedColor]}];
    [expectedNewLine addAttribute:NSBackgroundColorAttributeName value:[UIColor strongAddedColor] range:NSMakeRange(29, 1)];
    [expectedNewLine addAttribute:NSBackgroundColorAttributeName value:[UIColor strongAddedColor] range:NSMakeRange(70, 1)];

    JSDiff *diffObject = [[JSDiff alloc] initWithDeletedColor: [UIColor deletedColor] deletedWordColor: [UIColor strongDeletedColor] addedColor: [UIColor addedColor] addedWordColor: [UIColor strongAddedColor]];
    JSLineDiff *lineDiff = [diffObject diffWords:oldLine newLine:newLine];

    XCTAssertEqualObjects(lineDiff.oldLine, expectedOldLine);
    XCTAssertEqualObjects(lineDiff.newLine, expectedNewLine);
}

@end

File: JSDiffTests/JSDiffTests.swift
//
//  JSDiffTests.swift
//  JSDiffTests
//
//  Created by Piet Brauer on 09/04/16.
//  Copyright © 2016 nerdish by nature. All rights reserved.
//

import XCTest
import JSDiff

class JSDiffTests: XCTestCase {
    func testChangedNumbers() {
        let oldLine = "<a id=\"appIcon\" href=\"https://itunes.apple.com/us/app/git2go-git-client-you-always/id963577401?mt=8\"><img style=\"height:125px; width:125px\" src=\"{{ site.url }}/img/App Icon 2 - 120.svg\" alt=\"Git2Go Icon\"></a>"
        let newLine = "<a id=\"appIcon\" href=\"https://itunes.apple.com/us/app/git2go-enterprise-git-client/id1079716387?ls=1&mt=8\"><img style=\"height:125px; width:125px\" src=\"{{ site.url }}/img/App Icon 2 - 120.svg\" alt=\"Git2Go Icon\"></a>"
        let jsDiff = JSDiff(deletedColor: .deletedColor(), deletedWordColor: .strongDeletedColor(), addedColor: .addedColor(), addedWordColor: .strongAddedColor())

        let expectedOldLine = NSMutableAttributedString(string: oldLine, attributes: [.backgroundColor: UIColor.deletedColor()])
        expectedOldLine.addAttributes([.backgroundColor: UIColor.strongDeletedColor()], range: (oldLine as NSString).range(of: "-you-always"))
        expectedOldLine.addAttributes([.backgroundColor: UIColor.strongDeletedColor()], range: (oldLine as NSString).range(of: "id963577401"))

        let expectedNewLine = NSMutableAttributedString(string: newLine, attributes: [.backgroundColor: UIColor.addedColor()])
        expectedNewLine.addAttributes([.backgroundColor: UIColor.strongAddedColor()], range: (newLine as NSString).range(of: "enterprise-"))
        expectedNewLine.addAttributes([.backgroundColor: UIColor.strongAddedColor()], range: (newLine as NSString).range(of: "id1079716387"))
        expectedNewLine.addAttributes([.backgroundColor: UIColor.strongAddedColor()], range: (newLine as NSString).range(of: "ls=1&"))

        let result = jsDiff.diffWords(oldLine, newLine: newLine)
        XCTAssertEqual(result.oldLine, expectedOldLine)
        XCTAssertEqual(result.newLine, expectedNewLine)
    }
}
